File: matlab/unit/demo_fir.m
%  Running this demos least squares estimation 
%  of FIR model structure

clear; close all;  
global dm; if isempty(dm), clear global dm; dm=0; end
global dsp; if isempty(dsp), clear global dsp; dsp=1; end
global trans; 
if dsp>1, clc; echo on; end
OPT.dsp = dsp;

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%  Specify Experiment Conditions                    
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

T   = 0.1;    % Sampling period in seconds
var = 1e-4;   % Measurement noise variance

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%  Specify true system - ZOH impulse fresponse of cts defined system
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

bq = impulse(c2d(tf(1,real(poly([-1 -3+j -3-j]))),0.1,'zoh'));

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%  Simulate a data record                    
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

N       = 50*length(bq);
Z.u     = randn(1,N);
noise   = sqrt(var)*randn(size(Z.u));
Z.y     = filter(bq,1,Z.u)+noise; 

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%  Specify Model Structure                    
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

M.T     = T; 
M.type  = 'fir';
M.nB    = length(bq)-1;

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%  Estimate on basis of noise corrupted data
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

G = est(Z,M,OPT);

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%  Plot the results
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
if dsp,
	Gt.B = bq(:)';
	Gt.type = 'fir';
	Gt.w = G.w; 
	Gt.T = G.T;  
	Gt.disp.colour='b';
	Gt.disp.legend = 'True Response';
	G.disp.legend = [G.disp.legend, ' q operator'];
	Gt.disp.aux='magonly';
	Gt = startM(Z,Gt);
	
	showbode(Gt,G);
end

echo off;

if dm
 disp('  ')
 disp('---------------------------------------------------------------------')
 disp('  ')
 disp('You now have access to the MATLAB workspace so that you may examine')
 disp('the results of this simulation.  To return to the demos, type "dbcont"')
 disp(' ')
 keyboard; 
end;

File: src/uonidtoolbox/demo/_demo_fir.py

import uonidtoolbox as unit
import numpy as np
import scipy


def demo_fir(disp=1):

    OPT = unit.struct()
    OPT.dsp = disp

    # ====================================================
    # Specify Experiment Conditions
    # ====================================================
    T   = 0.5   # Sampling period in seconds
    var = 1e-4  # Measurement Noise Variance


    # ====================================================
    # Specify true system - ZOH impulse response of continuous defined system
    # ====================================================
    den     = np.real(np.poly([-1, -3+1j, -3-1j]))
    sysc    = scipy.signal.TransferFunction(1, den)
    sysd    = scipy.signal.cont2discrete((sysc.num, sysc.den), T, method='zoh')
    bq      = scipy.signal.dimpulse(sysd)[1][0].ravel()


    # ====================================================
    # Simulate a data record
    # ====================================================
    Z       = unit.struct()
    N       = 50*bq.shape[0]
    Z.u     = np.random.randn(1, N)
    noise   = np.sqrt(var)*np.random.randn(Z.u.size).reshape(Z.u.shape)
    Z.y     = scipy.signal.lfilter(bq, 1, Z.u) + noise


    # ====================================================
    # Specify Model Structures
    # ====================================================
    Mq          = unit.struct()
    Mq.nB       = bq.shape[0]-1
    Mq.T        = T
    Mq.type     = 'fir'


    # ====================================================
    # Estimate on basis of noise corrupted data
    # ====================================================
    Gq = unit.est(Z,Mq,OPT)


    # ====================================================
    # Plot the results
    # ====================================================
    if OPT.dsp:
        Gt              = unit.struct()
        Gt.disp         = unit.struct()
        Gt.B            = bq
        Gt.T            = T
        Gt.w            = Gq.w
        Gt.type         = 'fir'
        # Gt.colour       = 'b'
        Gt.disp.legend  = 'True Response'
        Gt.disp.legend  = Gt.disp.legend + ' q operator'
        
        unit.plotting.showbode([Gt, Gq])
    #endif

#endfunction

